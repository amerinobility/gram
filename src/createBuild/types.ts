import { GraphQLResolveInfo, GraphQLSchema, GraphQLType } from 'graphql'
import { IResolvers, ITypeDefinitions, MergeInfo } from 'graphql-tools'

import { GQLBUILDER } from '../types/constants'

import { AddObjectTypeConfig, createAddType } from './method-addType'

// Context == GraphQLContext ({ user: {â€¦}, db: DBConnection })
// BuildMode == BuildMode of the build ('user' | 'admin')

/// replaces the FieldResolver to have a ResultType as well
export type GraphQLFieldResolver<Source, Args, Context, Result> = (
  source: Source,
  args: Args,
  context: Context,
  info: GraphQLResolveInfo & { mergeInfo: MergeInfo },
) => Result

export type MaybeGenerated<Type, BuildMode> =
  | Type
  | BuildModeGenerator<BuildMode, Type>

export type MaybeGeneratedSimpleFieldType<
  BuildMode,
  Source,
  Context
> = MaybeGenerated<SimpleFieldType<Source, Context>, BuildMode>

export interface BaseResolvableConfig {
  args?: GQLRecord
}

export interface QueryResolvableConfig<
  Source,
  Context,
  Args = any,
  Result = any
> extends BaseResolvableConfig {
  resolver?: GraphQLFieldResolver<Source, Args, Context, Result>
}
export interface MutationResolvableConfig<
  Source,
  Context,
  Args = any,
  Result = any
> extends BaseResolvableConfig {
  resolver?: GraphQLFieldResolver<Source, Args, Context, Result>
}
export interface SubscriptionResolvableConfig<
  Source,
  Context,
  Args = any,
  Result = any
> extends BaseResolvableConfig {
  resolve?: GraphQLFieldResolver<never, Args, Context, Result>
  subscribe?: GraphQLFieldResolver<Source, Args, Context, AsyncIterator<Result>>
}
export type AddResolvable<BuildMode, Context> = ((
  typeName: 'Query',
) => AddQuery<BuildMode, Context>) &
  ((typeName: 'Mutation') => AddMutation<BuildMode, Context>) &
  ((typeName: 'Subscription') => AddSubscription<BuildMode, Context>)
export type AddQuery<BuildMode, Context> = <Source, Args = any, Result = any>(
  name: string,
  type: MaybeGeneratedSimpleFieldType<BuildMode, Source, Context>,
  config?: QueryResolvableConfig<Source, Context, Args, Result>,
) => void
export type AddSubscription<BuildMode, Context> = <
  Source,
  Args = any,
  Result = any
>(
  name: string,
  type: MaybeGeneratedSimpleFieldType<BuildMode, Source, Context>,
  config: SubscriptionResolvableConfig<Source, Context, Args, Result>,
) => void
export type AddMutation<BuildMode, Context> = <
  Source,
  Args = any,
  Result = any
>(
  name: string,
  type: MaybeGeneratedSimpleFieldType<BuildMode, Source, Context>,
  config: MutationResolvableConfig<Source, Context, Args, Result>,
) => void

export type AddResolvableConfig<Source, Context, Args = any, Result = any> =
  | QueryResolvableConfig<Source, Context, Args, Result>
  | SubscriptionResolvableConfig<Source, Context, Args, Result>
  | MutationResolvableConfig<Source, Context, Args, Result>

export interface TypeDefs {
  typeDefs: ITypeDefinitions
  resolvers: IResolvers
}
export interface Build<BuildMode, Context> {
  type: typeof GQLBUILDER
  buildMode: BuildMode
  addMutation: AddMutation<BuildMode, Context>
  addQuery: AddQuery<BuildMode, Context>
  addResolver: <Source>(
    base: string,
    name: string,
    resolver: AddResolvableConfig<Source, Context>,
  ) => void
  addSubscription: AddSubscription<BuildMode, Context>
  addType: ReturnType<typeof createAddType>
  extendType: <Source>(
    typeName: string,
    config: AddObjectTypeConfig<BuildMode, Source, Context>,
  ) => void
  isScalar: (type: string) => boolean
  isType: (type: string) => boolean
  toSchema: () => GraphQLSchema
  toTypeDefs: () => TypeDefs
  getState: () => any
}

export type GQLRecord = Record<string, string | GraphQLType>
export type Resolvers = IResolvers
export type BuildModeGenerator<BuildMode, Result> = (
  buildMode?: BuildMode,
) => Result
export type BuildModeArgsGenerator<BuildMode, Args, Result = void> = (
  buildModeGenerator: BuildModeGenerator<BuildMode, Args>,
) => Result
export interface FieldType<Source, Context, Args = any, Result = any> {
  args?: GQLRecord
  type: string | GraphQLType
  resolver?: GraphQLFieldResolver<Source, Args, Context, Result>
}
export type SimpleFieldType<Source, Context> =
  | string
  | GraphQLType
  | FieldType<Source, Context>
export type Fields<Source, Context> = Record<string, FieldType<Source, Context>>

export interface CreateableTypeConfig<Source, Context> {
  fields: Fields<Source, Context>
}

export type Resolvables = 'Query' | 'Mutation' | 'Subscription'
export type ResolvablesRecord<Source, Context> = Record<
  Resolvables,
  CreateableTypeConfig<Source, Context>
>
export interface ObjectTypesRecordConfig<Source, Context>
  extends CreateableTypeConfig<Source, Context> {
  interface?: string
}
// tslint:disable-next-line:no-empty-interface
export interface InterfaceTypesRecordConfig<Source, Context>
  extends CreateableTypeConfig<Source, Context> {}

export interface InputTypesRecordConfig<Source, Context>
  extends CreateableTypeConfig<Source, Context> {
  interface?: string
}
export interface EnumTypesRecordConfig {
  values: string[]
}
export interface CreateableTypesRecord<Source, Context> {
  scalar: string[]
  type: Record<string, ObjectTypesRecordConfig<Source, Context>>
  interface: Record<string, InterfaceTypesRecordConfig<Source, Context>>
  input: Record<string, InputTypesRecordConfig<Source, Context>>
  enum: Record<string, EnumTypesRecordConfig>
}
export type CreateableTypes<Source, Context> = keyof CreateableTypesRecord<
  Source,
  Context
>

// string === scalar
export type AnyCreateableTypeConfig<Source, Context> =
  | string
  | ObjectTypesRecordConfig<Source, Context>
  | InterfaceTypesRecordConfig<Source, Context>
  | InputTypesRecordConfig<Source, Context>
  | EnumTypesRecordConfig

export type AddResolver<Context> = <Source>(
  base: string,
  name: string,
  config: AddResolvableConfig<Source, Context>,
) => void
