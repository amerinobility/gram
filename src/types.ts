
import {
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLInputFieldConfigMap,
  GraphQLInputObjectType,
  GraphQLInterfaceType,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLType,
  Thunk,
} from 'graphql'

import { Names } from 'strategies/naming'

export type Fields = Thunk<GraphQLFieldConfigMap<any, any>>

export type DataType = 'create' | 'filter' | 'data' | 'page' | 'where' | 'order'

export interface ContextModel {
  addField: (name: string, field: GraphQLFieldConfig<any, any>) => void
  baseFilters: () => GraphQLInputFieldConfigMap
  dataFields: (type: DataType) => GraphQLInputFieldConfigMap | GraphQLFieldConfigMap<any, any>
  getFields: () => GraphQLFieldConfigMap<any, any>
  setInterface: () => void
  isInterface: () => boolean
  getListType: () => GraphQLType
  getPubSub: () => any
  getType: () => GraphQLType
  id: string
  name: string
  names: Names
  visibility: ModelVisibility
}

export interface Wrapped<Context> {
  getModel: (name: string) => ContextModel
  addModel: (name: string, model: ContextModel) => void
  context: Context
}

export type ContextFn<Context, Result = boolean> = (context: Wrapped<Context>) => Result
export type ContextMutator<Context> = (model: ContextModel, context: Wrapped<Context>) => void
export type ContextModelFieldFn<Type> = (contextModel: ContextModel) => Type

export type ModelType<Context> = GraphQLType | ModelBuilder<Context, any>

export interface SchemaBuilder<Context> {
  // creates a new Model and adds it to the schema
  // modelName = name of the Model in Graphql
  // contextFn = is the model visible in this context (default = true)
  model: <Type>(modelName: string, contextFn?: ContextFn<Context>) => ModelBuilder<Context, Type>
  interface: <Type>(interfaceName: string, contextFn?: ContextFn<Context>) => ModelBuilder<Context, Type>
  models: Record<string, ModelBuilder<Context, any>>
  build: (context: Context) => GraphQLSchema
}

export interface ModelBuilder<Context, Type> {
  attr: (attributeName: string, type: ModelType<Context>) => AttributeBuilder<Context, Type>
  build: (context: Wrapped<Context>) => ContextModel
  context: (contextMutation: ContextMutator<Context>) => this
  getInterfaces: () => string[]
  getListType: () => GraphQLType | ModelBuilder<Context, any>
  interface: (model: string) => this
  isInterface: () => boolean
  listType: (model: ModelBuilder<Context, any>) => this
  name: string
  setInterface: () => this
  setup: ContextFn<Context, void>
}

export interface AttributeBuilder<Context, Type> {
  name: string
  type: ContextFn<Context, GraphQLType | ModelBuilder<Context, Type>>
  nullable: boolean
  listType: boolean
  resolve: <Result>(resolver: (type: Type) => Result) => this
  isList: (isList?: boolean) => this
  isNotNullable: (isNotNullable?: boolean) => this
  build: ContextFn<Context, GraphQLFieldConfig<any, any>>
}

export interface ModelVisibility {
  createMutation: boolean
  deleteMutation: boolean
  findManyQuery: boolean
  findOneQuery: boolean
  updateMutation: boolean
  createSubscription: boolean
  updateSubscription: boolean
  deleteSubscription: boolean
}
