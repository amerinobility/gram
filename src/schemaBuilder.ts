import {
  GraphQLFieldConfigMap,
  GraphQLFieldResolver,
  GraphQLID,
  GraphQLInputFieldConfigMap,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  GraphQLType,
} from 'graphql'
import { forEach, reduce } from 'lodash'
import { createModelBuilder } from './modelBuilder'
import {
  ContextModel,
  ContextMutator,
  ListType,
  ModelBuilder,
  NodeType,
  PageData,
  SchemaBuilder,
  Service,
  Wrapped,
} from './types'

import { v4 as uuid } from 'uuid'
import {
  mutationFieldsReducer,
  queryFieldsReducer,
  subscriptionFieldsReducer,
} from './field-reducers'
import { SCHEMABUILDER } from './types/constants'

const wrapContext = <Context>(context: Context | null): Wrapped<Context> => {
  const models: Record<string, ContextModel<Context, any>> = {}
  return {
    id: uuid(),
    context,
    addModel: (name, model) => (models[name] = model),
    getModel: name => models[name],
  }
}

type BaseTypes = 'Page' | 'Node' | 'List'

export const createSchemaBuilder = <Context = any>(): SchemaBuilder<
  Context
> => {
  const addNodeAttrs = (model: ModelBuilder<Context, any>) => {
    model.attr('id', GraphQLID)
    model.attr('createdAt', GraphQLString)
    model.attr('updatedAt', GraphQLString)
    model.attr('deletedAt', GraphQLString)
    return model
  }

  const node = createModelBuilder<Context, NodeType>('Node', {}).setInterface()
  const page = createModelBuilder<Context, PageData>('Page', {})
  const list = createModelBuilder<Context, ListType<NodeType>>(
    'List',
    {},
  ).setInterface()

  addNodeAttrs(node)

  page.attr('page', GraphQLInt)
  page.attr('limit', GraphQLInt)
  page.attr('offset', GraphQLInt)

  list.attr('page', page)
  list.attr('nodes', node).isList()

  const models: Record<string, ModelBuilder<Context, any>> = {
    Node: node,
    Page: page,
    List: list,
  }

  const builder: SchemaBuilder<Context> = {
    type: SCHEMABUILDER,
    models,
    model: <Type>(modelName: string, service: Service<Type>) => {
      const model = createModelBuilder<Context, Type>(modelName, service || {})
      models[modelName] = model
      return addNodeAttrs(model.interface('Node'))
    },
    interface: <Type>(interfaceName: string, service: Service<Type>) =>
      builder.model<Type>(interfaceName, service).setInterface(),
    build: (context: Context | null = null) => {
      const wrapped = wrapContext<Context>(context)
      forEach(models, model => model.setup(wrapped))

      models.Node.build(wrapped)
      models.Page.build(wrapped)
      models.List.build(wrapped)

      const mutationFields: GraphQLFieldConfigMap<any, any> = reduce(
        models,
        mutationFieldsReducer(wrapped),
        {},
      )
      const queryFields: GraphQLFieldConfigMap<any, any> = reduce(
        models,
        queryFieldsReducer(wrapped),
        {},
      )
      const subscriptionFields: GraphQLFieldConfigMap<any, any> = reduce(
        models,
        subscriptionFieldsReducer(wrapped),
        {},
      )

      return new GraphQLSchema({
        query: new GraphQLObjectType({
          name: 'Query',
          fields: queryFields,
        }),
        mutation: new GraphQLObjectType({
          name: 'Mutation',
          fields: mutationFields,
        }),
        subscription: new GraphQLObjectType({
          name: 'Subscription',
          fields: subscriptionFields,
        }),
      })
    },
  }

  return builder
}
