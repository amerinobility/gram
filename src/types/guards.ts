import {
  AttributeBuilder,
  Builder,
  ModelBuilder,
  SchemaBuilder,
} from '../types'
import {
  ATTRIBUTEBUILDER,
  MODELBUILDER,
  SCHEMABUILDER,
} from '../types/constants'

export const isBuilder = (val: any): val is Builder =>
  val && typeof val.type === 'string'
export const isAttributeBuilder = <Context, Type, AttributeType>(
  val: any,
): val is AttributeBuilder<Context, Type, AttributeType> =>
  isBuilder(val) && val.type === ATTRIBUTEBUILDER
export const isModelBuilder = <Context, Type>(
  val: any,
): val is ModelBuilder<Context, Type> =>
  isBuilder(val) && val.type === MODELBUILDER
export const isSchemaBuilder = <Context>(
  val: any,
): val is SchemaBuilder<Context> =>
  isBuilder(val) && val.type === SCHEMABUILDER
