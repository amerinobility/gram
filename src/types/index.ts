import {
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLFieldResolver,
  GraphQLInputFieldConfigMap,
  GraphQLInputObjectType,
  GraphQLInterfaceType,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLType,
  Thunk,
} from 'graphql'

import { Names } from '../strategies/naming'
import {
  ATTRIBUTEBUILDER,
  MODELBUILDER,
  SCHEMABUILDER,
} from '../types/constants'
import { Service } from '../types/service'
export * from '../types/service'

export interface Builder {
  type: string
}

export type Fields = Thunk<GraphQLFieldConfigMap<any, any>>

export type DataType =
  | 'create'
  | 'filter'
  | 'data'
  | 'list'
  | 'page'
  | 'where'
  | 'order'

export type FieldTypes = 'GraphQL' | 'Model' | 'All'
export type GenericGraphQLType = 'Date' | 'Upload' | 'JSON'

export interface ContextModel<Context, Type> {
  addField: <AttributeType>(
    field: AttributeBuilder<Context, Type, AttributeType>,
  ) => void
  baseFilters: () => GraphQLInputFieldConfigMap
  context: Wrapped<Context>
  dataFields: (
    type: DataType,
  ) => GraphQLInputFieldConfigMap | GraphQLFieldConfigMap<any, any>
  getFields: () => Array<AttributeBuilder<Context, Type, any>>
  getListType: () => GraphQLType
  getPubSub: () => any
  getType: () => GraphQLType
  id: string
  isInterface: () => boolean
  name: string
  names: Names
  service: Service<Type>
  setInterface: () => void
  visibility: ModelVisibility
}

export interface Wrapped<Context> {
  id: string
  getModel: <Type>(name: string) => ContextModel<Context, Type>
  addModel: <Type>(name: string, model: ContextModel<Context, Type>) => void
  context: Context | null
  getGenericType: (key: string) => GraphQLType
}

export type ContextFn<Context, Result = boolean> = (
  context: Wrapped<Context>,
) => Result
export type ContextMutator<Context, Type> = (
  model: ContextModel<Context, Type>,
  context: Wrapped<Context>,
) => void
export type ContextModelFieldFn<Type> = <Context>(
  contextModel: ContextModel<Context, Type>,
) => Type

export type ModelType<Context> = GraphQLType | ContextModel<Context, any>

export interface SchemaBuilder<Context> extends Builder {
  build: (context?: Context) => GraphQLSchema
  interface: <Type>(
    interfaceName: string,
    service?: Service<Type>,
  ) => ModelBuilder<Context, Type>
  model: <Type>(
    modelName: string,
    service?: Service<Type>,
  ) => ModelBuilder<Context, Type>
  models: Record<string, ModelBuilder<Context, any>>
  type: typeof SCHEMABUILDER
  setGenericType: <Type extends GraphQLType>(
    key: GenericGraphQLType,
    type: Type,
  ) => Type
  getGenericType: (key: GenericGraphQLType) => GraphQLType
}

export interface ModelBuilder<Context, Type> extends Builder {
  attr: <AttributeType>(
    attributeName: string,
    type:
      | ModelType<Context>
      | ModelBuilder<Context, any>
      | ContextFn<Context, GraphQLType>,
  ) => AttributeBuilder<Context, Type, AttributeType>
  build: (context: Wrapped<Context>) => ContextModel<Context, Type>
  context: (contextMutation: ContextMutator<Context, Type>) => this
  getInterfaces: () => string[]
  getListType: () => GraphQLType | ModelBuilder<Context, any>
  interface: (model: string) => this
  isInterface: () => boolean
  listType: (model: ModelBuilder<Context, any>) => this
  name: string
  setInterface: () => this
  setup: ContextFn<Context, void>
  type: typeof MODELBUILDER
}

export interface AttributeBuilder<Context, Type, AttributeType>
  extends Builder {
  name: string
  field: ContextFn<Context, GraphQLType | ContextModel<Context, Type>>
  nullable: boolean
  listType: boolean
  resolve: (resolver: GraphQLFieldResolver<Type, Context>) => this
  isList: (isList?: boolean) => this
  isNotNullable: (isNotNullable?: boolean) => this
  // contextType: ContextFn<Context, ModelType<Context>>
  build: ContextFn<Context, GraphQLFieldConfig<any, any>>
  type: typeof ATTRIBUTEBUILDER
}

export interface ModelVisibility {
  createMutation: boolean
  deleteMutation: boolean
  findManyQuery: boolean
  findOneQuery: boolean
  updateMutation: boolean
  createSubscription: boolean
  updateSubscription: boolean
  deleteSubscription: boolean
}
