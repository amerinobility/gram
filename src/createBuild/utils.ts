import { GraphQLType, isType } from 'graphql'
import { isEmpty, map, reduce } from 'lodash'
import { isBuildModeGenerator } from './guards'
import {
  BuildModeGenerator,
  Fields,
  FieldType,
  GQLRecord,
  SimpleFieldType,
} from './types'

export const typeToString = (type: string | GraphQLType) =>
  isType(type) ? type.toString() : type

export const isListType = (type: string) =>
  type[0] === '[' && type[type.length - 1] === ']'
export const isNonNullableList = (type: string) =>
  type[0] === '[' &&
  type[type.length - 2] === ']' &&
  type[type.length - 1] === '!'
export const isNullable = (type: string) => type[type.length - 1] !== '!'
export const getParentType = (type: string) =>
  !isNullable(type)
    ? getParentType(type.substr(0, type.length - 1))
    : isListType(type)
    ? getParentType(type.substr(1, type.length - 2))
    : type

export const nullable = (type: string) =>
  isNullable(type) ? type : type.substr(0, type.length - 1)

export const nonNull = (type: string) => `${getParentType(type)}!`
export const list = (type: string, nullable = false) =>
  nullable ? `[${nonNull(type)}]` : `[${nonNull(type)}]!`

export const reduceRecord = <Source, Context>(
  fields: Fields<Source, Context>,
) =>
  reduce(
    fields,
    (list, { args = {}, type }, name) => {
      const parameters = isEmpty(args)
        ? ''
        : `(${map(args, (type, arg) => `${arg}: ${type}`)})`
      list.push(`${name}${parameters}: ${type}`)
      return list
    },
    [],
  ).join('\n    ')

export const createBuildModeResolver = <BuildMode>(buildMode: BuildMode) => <
  Result
>(
  data: Result | BuildModeGenerator<BuildMode, Result>,
) => (isBuildModeGenerator<BuildMode, Result>(data) ? data(buildMode) : data)

export const simpleFieldTypeToFieldType = <Source, Context>(
  simpleType: SimpleFieldType<Source, Context>,
): FieldType<Source, Context> => {
  if (isType(simpleType)) return { type: simpleType.toString() }
  if (typeof simpleType === 'string') return { type: simpleType }
  return simpleType
}

export const convertSimpleFieldsToFields = <Source, Context>(
  simpleFields: Record<string, SimpleFieldType<Source, Context>>,
): Fields<Source, Context> =>
  reduce(
    simpleFields,
    (record, simpleField, name) => {
      record[name] = simpleFieldTypeToFieldType(simpleField)
      return record
    },
    {},
  )
