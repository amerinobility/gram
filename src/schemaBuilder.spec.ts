
import { graphql, GraphQLID, GraphQLInt, GraphQLList, GraphQLString, printSchema } from 'graphql'

import { createSchemaBuilder } from './schemaBuilder'
import { ModelBuilder, SchemaBuilder } from './types'

import { toList } from 'utils'

describe('schemaBuilder', () => {
  let builder: SchemaBuilder<number>

  beforeEach(() => {
    builder = createSchemaBuilder<number>()
  })

  it('should have basic methods', () => {
    expect(builder).toHaveProperty('model')
    expect(typeof builder.model).toBe('function')
  })

  it('should render an empty Schema', () => {
    expect(printSchema(builder.build(0))).toMatchSnapshot()
  })
})

// interfaces
interface NodeType {
  id: string
  createdAt: Date
  updatedAt: Date
  deletedAt: Date | null
}
interface Page {
  limit: number
  offset: number
  page: number
}
interface Paged<Type> {
  page: Page
  nodes: Type[]
}

interface Account {
  id: string
  name: string
  user: User
}
interface User {
  id: string
  name: string
  accounts: Paged<Account>
}

describe.only('example', () => {
  // services
  let db = {}
  const Accounts = { findAll: () => [] }
  const Users = { find: () => null }

  let builder: SchemaBuilder<number>

  beforeAll(() => {
    db = {}
    builder = createSchemaBuilder<number>()
  })

  const addNodeAttributes = <Type>(node: ModelBuilder<number, Type>) => {
    node.attr('id', GraphQLID)
    node.attr('createdAt', GraphQLString)
    node.attr('updatedAt', GraphQLString)
    node.attr('deletedAt', GraphQLString)
    return node
  }

  const addListAttributes = <ListType, ModelType>(
    list: ModelBuilder<number, ListType>,
    model: ModelBuilder<number, ModelType>,
  ) => {
    list.attr('page', builder.models.Page)
    list.attr('nodes', model).isList()
    return list
  }

  const hide = <Type>(model: ModelBuilder<number, Type>) => {
    const hidden = {
      createMutation: false,
      deleteMutation: false,
      findManyQuery: false,
      findOneQuery: false,
      updateMutation: false,
    }
    model.context(model => { model.visibility = { ...model.visibility, ...hidden } })
    return model
  }

  const createListOf = <Type>(name: string, model: ModelBuilder<number, Type>) => {
    if(builder.models[name])
      return builder.models[name]
    const list = hide(addListAttributes(builder.model<Type>(name).interface('List'), model))
    model.listType(list)
    return list
  }

  it('should add User to the Schema', () => {
    const user = builder.model<User>('User')
    user.attr('name', GraphQLString)

    expect(printSchema(builder.build(0))).toMatchSnapshot()
  })

  it('should accept an interface type', () => {
    const node = hide(addNodeAttributes(builder.interface<NodeType>('Node')))
    const page = hide(builder.model<NodeType>('Page'))
    const list = hide(addListAttributes(builder.interface<NodeType>('List'), node))

    page.attr('page', GraphQLInt)
    page.attr('limit', GraphQLInt)
    page.attr('offset', GraphQLInt)

    createListOf('Users', addNodeAttributes(builder.models.User.interface('Node')))

    expect(printSchema(builder.build(0))).toMatchSnapshot()
  })

  it('should add account to the model', () => {
    const account = addNodeAttributes(builder.model<Account>('Account').interface('Node'))
    const accounts = createListOf('Accounts', account)

    // will add input types for "STRING"
    account.attr('name', GraphQLString)
    // will add input types for "REFERENCE"
    account
      .attr('user', builder.models.User)
      .isNotNullable()
    // add accounts to the user model
    builder.models.User
      .attr('accounts', accounts)

    expect(printSchema(builder.build(0))).toMatchSnapshot()
  })

  it('should be able to use the schema', async () => {
    const schema = builder.build(0)

    const query = `{
      getUser {
        name
      }
    }`

    console.log(printSchema(schema))
    const { data, errors } = await graphql(schema, query, null)
    if(errors) {
      console.log(errors[0])
      console.log(errors[0].locations)
    }
    console.log(data)
  })
})
