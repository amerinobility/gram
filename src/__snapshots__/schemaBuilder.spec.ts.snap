// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example should add User to the Schema 1`] = `
"input CreateUserData {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  friends: UserWhere
}

interface List {
  page: Page
  nodes: [Node!]!
}

type Mutation {
  createUser(data: CreateUserData!): User
  updateUser(data: UpdateUserData!, where: UserWhere!): [User!]!
  deleteUsers(where: UserWhere!): [User!]!
}

interface Node {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getUser(where: UserWhere!, order: UserSortOrder): User
  getUsers(order: UserSortOrder, page: UserPage, where: UserWhere!): Users
}

type Subscription {
  onCreateUser: User!
  onUpdateUser: [User!]!
  onDeleteUsers: [User!]!
}

input UpdateUserData {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  friends: UserWhere
}

type User implements Node {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  friends: [User!]!
}

input UserFilter {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  createdAt: String
  createdAt_not: String
  createdAt_in: [String!]
  createdAt_not_in: [String!]
  updatedAt: String
  updatedAt_not: String
  updatedAt_in: [String!]
  updatedAt_not_in: [String!]
  deletedAt: String
  deletedAt_not: String
  deletedAt_in: [String!]
  deletedAt_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
}

input UserPage {
  limit: Int
  offset: Int
}

type Users implements List {
  page: Page
  nodes: [User!]!
}

enum UserSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  friends_ASC
  friends_DESC
}

input UserWhere {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  createdAt: String
  createdAt_not: String
  createdAt_in: [String!]
  createdAt_not_in: [String!]
  updatedAt: String
  updatedAt_not: String
  updatedAt_in: [String!]
  updatedAt_not_in: [String!]
  deletedAt: String
  deletedAt_not: String
  deletedAt_in: [String!]
  deletedAt_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
}
"
`;

exports[`example should add account to the model 1`] = `
"type Account implements Node {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  amount: Float
  user: User!
}

input AccountFilter {
  AND: [AccountFilter!]
  OR: [AccountFilter!]
  NOT: AccountFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  createdAt: String
  createdAt_not: String
  createdAt_in: [String!]
  createdAt_not_in: [String!]
  updatedAt: String
  updatedAt_not: String
  updatedAt_in: [String!]
  updatedAt_not_in: [String!]
  deletedAt: String
  deletedAt_not: String
  deletedAt_in: [String!]
  deletedAt_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  amount: Float
  amount_not: Float
  amount_gt: Float
  amount_lt: Float
}

input AccountPage {
  limit: Int
  offset: Int
}

type Accounts implements List {
  page: Page
  nodes: [Account!]!
}

enum AccountSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  amount_ASC
  amount_DESC
  user_ASC
  user_DESC
}

input AccountWhere {
  AND: [AccountFilter!]
  OR: [AccountFilter!]
  NOT: AccountFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  createdAt: String
  createdAt_not: String
  createdAt_in: [String!]
  createdAt_not_in: [String!]
  updatedAt: String
  updatedAt_not: String
  updatedAt_in: [String!]
  updatedAt_not_in: [String!]
  deletedAt: String
  deletedAt_not: String
  deletedAt_in: [String!]
  deletedAt_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  amount: Float
  amount_not: Float
  amount_gt: Float
  amount_lt: Float
}

input CreateAccountData {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  amount: Float
  user: UserWhere!
}

input CreateUserData {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  friends: UserWhere
  accounts: AccountWhere
}

interface List {
  page: Page
  nodes: [Node!]!
}

type Mutation {
  createUser(data: CreateUserData!): User
  updateUser(data: UpdateUserData!, where: UserWhere!): [User!]!
  deleteUsers(where: UserWhere!): [User!]!
  createAccount(data: CreateAccountData!): Account
  updateAccount(data: UpdateAccountData!, where: AccountWhere!): [Account!]!
  deleteAccounts(where: AccountWhere!): [Account!]!
}

interface Node {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getUser(where: UserWhere!, order: UserSortOrder): User
  getUsers(order: UserSortOrder, page: UserPage, where: UserWhere!): Users
  getAccount(where: AccountWhere!, order: AccountSortOrder): Account
  getAccounts(order: AccountSortOrder, page: AccountPage, where: AccountWhere!): Accounts
}

type Subscription {
  onCreateUser: User!
  onUpdateUser: [User!]!
  onDeleteUsers: [User!]!
  onCreateAccount: Account!
  onUpdateAccount: [Account!]!
  onDeleteAccounts: [Account!]!
}

input UpdateAccountData {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  amount: Float
  user: UserWhere
}

input UpdateUserData {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  friends: UserWhere
  accounts: AccountWhere
}

type User implements Node {
  id: ID
  createdAt: String
  updatedAt: String
  deletedAt: String
  name: String
  friends: [User!]!
  accounts: [Account!]!
}

input UserFilter {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  createdAt: String
  createdAt_not: String
  createdAt_in: [String!]
  createdAt_not_in: [String!]
  updatedAt: String
  updatedAt_not: String
  updatedAt_in: [String!]
  updatedAt_not_in: [String!]
  deletedAt: String
  deletedAt_not: String
  deletedAt_in: [String!]
  deletedAt_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
}

input UserPage {
  limit: Int
  offset: Int
}

type Users implements List {
  page: Page
  nodes: [User!]!
}

enum UserSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  friends_ASC
  friends_DESC
  accounts_ASC
  accounts_DESC
}

input UserWhere {
  AND: [UserFilter!]
  OR: [UserFilter!]
  NOT: UserFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  createdAt: String
  createdAt_not: String
  createdAt_in: [String!]
  createdAt_not_in: [String!]
  updatedAt: String
  updatedAt_not: String
  updatedAt_in: [String!]
  updatedAt_not_in: [String!]
  deletedAt: String
  deletedAt_not: String
  deletedAt_in: [String!]
  deletedAt_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
}
"
`;

exports[`example should be able read from the schema 1`] = `
Object {
  "getUser": Object {
    "name": "test",
  },
}
`;

exports[`example should find multiple things 1`] = `
Object {
  "getUsers": Object {
    "nodes": Array [
      Object {
        "name": "test",
      },
      Object {
        "name": "New Test-User",
      },
    ],
  },
}
`;

exports[`schemaBuilder should render an empty Schema 1`] = `
"type Mutation

type Query

type Subscription
"
`;
